### PR TITLE
Add Socket::IPAddress#private?

### PR DESCRIPTION
**Pull request by jgillich**
Ported over from the ruby method with the same name.

### CODE REVIEW

File: src/socket/address.cr [hunk 1]
```diff
@@ -238,6 +238,22 @@ class Socket
       end
     end
 
+    # Returns `true` if this IP is a private address.
+    #
+    # IPv4 addresses in 10.0.0.0/8, 172.16.0.0/12 and 192.168.0.0/16 as defined in RFC 1918
+    # and IPv6 Unique Local Addresses in fc00::/7 as defined in RFC 4193 are considered private.
```
Comment (straight-shoota) [inline src/socket/address.cr:244]:
```suggestion
    # IPv4 addresses in `10.0.0.0/8`, `172.16.0.0/12` and `192.168.0.0/16` as defined in [RFC 1918](https://tools.ietf.org/html/rfc1918)
    # and IPv6 Unique Local Addresses in `fc00::/7` as defined in [RFC 4193](https://tools.ietf.org/html/rfc4193) are considered private.
```

Comment (jgillich) [reply src/socket/address.cr]: Applied, thanks

**Comment by wontruefree** [comment]
I really like this method ❤️

**Review by straight-shoota — APPROVED** [review]

File: spec/compiler/crystal/tools/doc/doc_renderer_spec.cr [hunk 2]
```diff
@@ -275,5 +275,14 @@ describe Doc::Markdown::DocRenderer do
...
     end
+
+    it "does not break when referencing lib type (#9928)" do
+      program = semantic("lib LibFoo; BAR = 0; end", wants_doc: true).program
+      generator = Doc::Generator.new(program, [""])
+
+      # TODO: There should not be a link to LibFoo::Bar in the first place
+      # because LibFoo is undocumented
+      assert_code_link(generator.type(program), "LibFoo::BAR", %(<a href="LibFoo.html#BAR">LibFoo::BAR</a>))
+    end
```
Comment (Sija) [inline spec/compiler/crystal/tools/doc/doc_renderer_spec.cr:286]: @straight-shoota Looks like a botched merge.

Comment (straight-shoota) [reply spec/compiler/crystal/tools/doc/doc_renderer_spec.cr:286]: Oh, that's messed up

Comment (jhass) [reply spec/compiler/crystal/tools/doc/doc_renderer_spec.cr:286]: Let's throw all that out with a `git rebase -i`, leaving only the original commit.

Comment (straight-shoota) [reply spec/compiler/crystal/tools/doc/doc_renderer_spec.cr]: Sorry for the mess. I made a hard reset and a clean merge of master.

File: src/socket/address.cr [hunk 3]
```diff
@@ -224,6 +224,21 @@ class Socket
...
+    # Returns `true` if this IP is a private address.
+    #
+    # IPv4 addresses in `10.0.0.0/8`, `172.16.0.0/12` and `192.168.0.0/16` as defined in [RFC 1918](https://tools.ietf.org/html/rfc1918)
+    # and IPv6 Unique Local Addresses in `fc00::/7` as defined in [RFC 4193](https://tools.ietf.org/html/rfc4193) are considered private.
+    def private? : Bool
+      case addr = @addr
+      in LibC::InAddr
+        addr.s_addr & 0x000000ff_u32 == 0x00000000a_u32 ||     # 10.0.0.0/8
+          addr.s_addr & 0x000000f0ff_u32 == 0x0000010ac_u32 || # 172.16.0.0/12
+          addr.s_addr & 0x000000ffff_u32 == 0x0000a8c0_u32     # 192.168.0.0/16
```
Comment (Sija) [inline src/socket/address.cr:236]: Small optimization:

```suggestion
        s_addr = addr.s_addr
        s_addr & 0x000000ff_u32 == 0x00000000a_u32 ||     # 10.0.0.0/8
          s_addr & 0x000000f0ff_u32 == 0x0000010ac_u32 || # 172.16.0.0/12
          s_addr & 0x000000ffff_u32 == 0x0000a8c0_u32     # 192.168.0.0/16
```

**Review by straight-shoota — APPROVED** [review]

**Comment by bcardiff** [comment]
This needs a rebase / conflict merge before merging. @jgillich if you prefer to hand off this, since it's been a while, please drop a comment.

**Comment by jgillich** [comment]
@bcardiff rebased

**Pull request merged**